### PR TITLE
optimize mutable.TreeSet#{max, min}

### PR DESCRIPTION
**Pull request by vadzim-marchanka**
Additional fix of mutable.TreeSet for https://github.com/scala/bug/issues/11598

### CODE REVIEW

**Comment by SethTisue** [comment]
Is it a 2.12-only issue, or does 2.13 have the problem as well?

**Comment by vadzim-marchanka** [comment]
@SethTisue thank you for the review. It is a 2.12-only issue because 2.13 has SortedSetOps with overridden methods.

File: src/library/scala/collection/mutable/TreeSet.scala [hunk 1]
```diff
@@ -104,6 +104,13 @@ sealed class TreeSet[A] private (tree: RB.Tree[A, Null])(implicit val ordering:
   override def last = RB.maxKey(tree).get
   override def lastOption = RB.maxKey(tree)
 
+
+  override def min[B >: A](implicit cmp: Ordering[B]): A =
+    if ((cmp eq ordering) && nonEmpty) head else super.min(cmp)
```
Comment (NthPortal) [inline src/library/scala/collection/mutable/TreeSet.scala]: none of the instances provided by the 2.12 standard library override `equals` (yet?), but there is not in principle a reason someone could not implement it for their own instance(s). consequently, using `==` (here and below) might be more appropriate. on the other hand, maybe it's not worth the effort

**Review by NthPortal — APPROVED** [review]
there's a minor tweak that I think would be a slight improvement, but honestly it borders on pedantry.

anyway, LGTM

File: src/library/scala/collection/mutable/TreeSet.scala (hunk 1 above)
Comment (vadzim-marchanka) [reply src/library/scala/collection/mutable/TreeSet.scala]: Thank you for the review. It seems like it's not worth the effort for me.  It is my first PR here 😊  and I do not want to complicate it. Could you please elaborate on your idea about overriding `equals` for own instance? What do you mean by "own instance" ?

Comment (NthPortal) [reply src/library/scala/collection/mutable/TreeSet.scala]: "instance" was the wrong word. I meant implementation. e.g. for `case class Foo[A](a: A, int: Int)`, you could implement an `Ordering` that overrides `equals`, as the implementations in 2.13 do (see https://github.com/scala/scala/blob/v2.13.8/src/library/scala/math/Ordering.scala#L281-L285 for example)

Comment (vadzim-marchanka) [reply src/library/scala/collection/mutable/TreeSet.scala]: Thank you for the explanation. I got it. I see one corner case. If somebody overrode `equals` in their own `Ordering` implementation in a wrong way (two different orderings are equal) then replacement `eq` by `==` will bring new behavior and bug to this person. Is it fine to introduce such changes?

Comment (NthPortal) [reply src/library/scala/collection/mutable/TreeSet.scala]: I mean, if someone implements `Ordering#compare` invalidly, everything will also be broken. we generally operate on the assumption that the user's code is not nonsense

Comment (vadzim-marchanka) [reply src/library/scala/collection/mutable/TreeSet.scala]: I have updated the PR (including immutable.TreeSet)

**Review by OndrejSpanel — APPROVED** [review]

**Comment by SethTisue** [comment]
Thank you Vadzim and reviewers!

**Pull request merged**